tap_freshdesk/streams.py

```diff
@@ -0,0 +1,315 @@
+        for child, bm in max_child_bms.items():
+            singer.write_bookmark(state, child, "updated_at", bm)
+
+class Tickets(DateFilteredStream):
+    tap_stream_id = 'tickets'
+    key_properties = ['id']
+    replication_keys = ['updated_at']
+    replication_method = 'INCREMENTAL'
+    path = 'tickets'
+    children = ['conversations', 'satisfaction_ratings', 'time_entries']
+    id_key = 'id'
+    date_filter = 'updated_since'
+    params = {
+        "per_page": PAGE_SIZE,
+        'order_by': replication_keys[0],
+        'order_type': "asc",
+        'include': "requester,company,stats"
+    }
+    filter_keyword = 'filter'
+    filters = [None, 'deleted', 'spam']
+
+class Contacts(DateFilteredStream):
+    tap_stream_id = 'contacts'
+    key_properties = ['id']
+    replication_keys = ['updated_at']
+    replication_method = 'INCREMENTAL'
+    path = 'contacts'
+    id_key = 'id'
+    date_filter = '_updated_since'
+    filter_keyword = 'state'
+    filters = [None, 'deleted', 'blocked']
+
+
+class ChildStream(Stream):
+
+    def sync_obj(self, state, start_date, client, catalog, selected_streams, streams_to_sync, predefined_filter=None):
+        """
+        The child stream sync_obj() method to sync the child records
+        """
+        params = copy.deepcopy(self.params)
+        # Build the url for the request
+        full_url = self.build_url(client.base_url, self.parent_id)
+        # Get the min bookmark from the parent and the child streams
+        min_bookmark = get_min_bookmark(self.tap_stream_id, streams_to_sync, start_date, state, self.replication_keys[0], None)
+        max_bookmark = min_bookmark
+        params['page'] = 1
+        self.paginate = True
+
+        LOGGER.info("Syncing %s from %s", self.tap_stream_id, min_bookmark)
+        # Paginate through the records
+        while self.paginate:
+            data = client.request(full_url, params)
+            self.paginate = len(data) >= PAGE_SIZE
+            params['page'] += 1
+            # Write the records based on the bookmark and return the max_bookmark for the page
+            bookmark, _ = self.write_records(catalog, state, selected_streams, start_date, data, max_bookmark, client, streams_to_sync, None)
+            max_bookmark = max(max_bookmark, bookmark)
+        return max_bookmark
+
+class Conversations(ChildStream):
+    tap_stream_id = 'conversations'
+    key_properties = ['id']
+    replication_keys = ['updated_at']
+    replication_method = 'INCREMENTAL'
+    path = 'tickets/{}/conversations'
+    parent = 'tickets'
+
+
+class SatisfactionRatings(ChildStream):
+    tap_stream_id = 'satisfaction_ratings'
+    key_properties = ['id']
+    replication_keys = ['updated_at']
+    replication_method = 'INCREMENTAL'
+    path = 'tickets/{}/satisfaction_ratings'
+    parent = 'tickets'
+
+class TimeEntries(ChildStream):
+    tap_stream_id = 'time_entries'
+    key_properties = ['id']
```

Comment:
The `key_properties` is `id` for all the streams so just set the value at the base class and remove it from all the stream classes.

Reply:
Moved `key_properties` to the base class.